app/controllers/api_controller.rb

```diff
@@ -1,96 +1,17 @@
-            cad: product.your_price_net(account, :cad, :unit)
-          },
-          piece: {
-            usd: product.your_price_net(account, :usd, :piece),
-            cad: product.your_price_net(account, :cad, :piece)
-          },
-          halfpiece: {
-            usd: product.your_price_net(account, :usd, :halfpiece),
-            cad: product.your_price_net(account, :cad, :halfpiece)
-          },
-          customer: {
-            unit: product.order_price__net(current_user, :unit),
-            piece: product.order_price__net(current_user, :piece),
-            halfpiece: product.order_price__net(current_user, :halfpiece)
-          }
-        }
-      },
-      stock: product.stock(current_user)&.merge(unit: product.measured_unit)
-    }
-
-    output[:uom_display_text] = product.measured_unit["uom_display_text"] if product.measured_unit["uom_display_text"].present?
-
-    if product.wallcovering?
-      output[:measured_unit] = product.measured_unit["long"]["singular"].downcase
-      output[:order_increment] = product.wallcovering_data.average_bolt.to_f
-    end
-
-    output[:pricing].delete(:net) unless current_user.has_role?(:view_net_pricing_usd) || current_user.has_role?(:view_net_pricing_cad)
-    output.delete(:stock) unless current_user.has_role?(:view_stock)
-
-    @product = product
-    @stock_output = output[:stock].deep_symbolize_keys
-
-    @stock_total = @stock_output[:current][:total]
-    @stock_unit = @stock_output[:unit][:long][@stock_total == 1 ? :singular : :plural]
-
-    @stock_total_reserved = @stock_output[:total_reserved]
-    @stock_total_reserved_unit = @stock_output[:unit][:long][@stock_total_reserved == 1 ? :singular : :plural]
-
-    @has_stock = @stock_total.present? && @stock_total.to_f.positive?
-    @has_availability = @stock_output[:availability].present?
-    @incoming_stock = @stock_output[:expected]
-
-    # Get all bolts and separate them into "regular bolts" and "small cuts"
-    @bolts = @stock_output[:current][:bolts]
-
-    # Small cuts applies to fabric sold by the yard, excluding Clarencehouse
-    if !clarencehouse? && @product.fabric? && @product.measured_unit["long"]["singular"] == "Yard"
-      @regular_bolts, @small_cuts = @bolts.partition { |_bolt, lot| lot[:quantity].to_f >= 5 } if @bolts.present?
-    else
-      @regular_bolts = @bolts
-      @small_cuts = nil
-    end
-
-    # Group and sort by dye lot number
-    @regular_bolts_dye_lots = @regular_bolts.group_by { |_bolt, lot| lot[:dye_lot] } if @regular_bolts.present?
-    @small_cuts_dye_lots = @small_cuts.group_by { |_bolt, lot| lot[:dye_lot] } if @small_cuts.present?
+    ApiHelper.calculate(Product.find(params[:sku]), current_account, current_user, clarencehouse?)
+    build_view_attributes
   rescue => e
     render json: { success: false, error: e.message }, status: :unprocessable_entity
   end
+
+  private
+
+  def build_view_attributes
+    ApiHelper.class_variables.map { |cv| eval("#{cv.to_s[1..-1]} = #{cv}") }
```

Comment:
This is what I love about getting a fresh set of eyes on a problem. `instance_variable_set` would be a way better approach than `eval`. I'm sure you're aware there are a few reasons using `eval` is pretty sketchy. Depending on how it's used, it can be a gift to a hacker, or a gift to a developer. So we've got to be extra careful no user inputs could be passed to this. In the way I've used it, that concern isn't there. However, with `instance_variable_set` && `class_variable_set` not only is that concern gone, but the intention becomes much more clear. With that the code becomes more readable and maintainable. So gold star for that suggestion! ⭐️

As slick as `eval` wasn't and `instance_variable_set` could have been, that doesn't really address the 2nd point you hit. I think if we take a step back and ask ourselves "_**why**_  do we have to solve this problem?" we would find that we've deviated from the MVC paradigm. Not having a model is causing us to have to fabricate solutions that ultimately don't need to exist.

You can see in my notes, I'm hinting at this larger conversation that needs to take place for the stock and pricing api feature. Luckily many rails apps have felt this pain. So much so, that `ActiveModel::API` was born. You can initialize objects that move just like active record objects (from a view perspective) which was ultimately the driving force of my hacky workaround. I'll quote directly from the opening remarks of [active model](https://guides.rubyonrails.org/active_model_basics.html)
> ActiveModel::API adds the ability for a class to work with Action Pack and Action View right out of the box.
> 
> ```ruby
> class EmailContact
>   include ActiveModel::API
> 
>   attr_accessor :name, :email, :message
>   validates :name, :email, :message, presence: true
> 
>   def deliver
>     if valid?
>       # deliver email
>     end
>   end
> end
> ```
> 
> When including ActiveModel::API you get some features like:
> 
> - model name introspection
> - conversions
> - translations
> - validations
> 
> It also gives you the ability to initialize an object with a hash of attributes, much like any Active Record object.
> 
> ```ruby
> irb> email_contact = EmailContact.new(name: 'David', email: 'david@example.com', message: 'Hello World')
> irb> email_contact.name
> => "David"
> irb> email_contact.email
> => "david@example.com"
> irb> email_contact.valid?
> => true
> irb> email_contact.persisted?
> => false
> ```
> 
> Any class that includes ActiveModel::API can be used with form_with, render and any other Action View helper methods, just like Active Record objects.

Obviously a larger refactor would need to take place for this and we would probably want to discuss how much of this of this could/should live within `Products`